doc/mgmt/gnmi/gnsi.md

```diff
@@ -0,0 +1,383 @@
+  - gNMI Server
+- What are the repositories that would be changed?
+  - [sonic-gnmi](https://github.com/sonic-net/sonic-gnmi)
+  - [sonic-mgmt-common](https://github.com/sonic-net/sonic-mgmt-common)
+- Module/sub-module interfaces and dependencies.
+  - Add some flags to the gNMI module that point to default PKI elements.
+  - Add flags to point to metadata files for each service
+- SWSS and Syncd changes in detail
+  - N/A
+- DB and Schema changes (APP_DB, ASIC_DB, COUNTERS_DB, LOGLEVEL_DB, CONFIG_DB, STATE_DB)
+  - Adds entries to STATE_DB for state and freshness of policies. See [Openconfig Models](https://github.com/openconfig/public/tree/master/release/models/gnsi)
+- Sequence diagram if required.
+ See [Certz.Rotate](#certzrotate)
+- Linux dependencies and interface
+  - N/A
+- Warm reboot requirements/dependencies
+  - None
+- Fastboot requirements/dependencies
+  - None
+- Scalability and performance requirements/impact
+  - There is a linear requirement for the numbers of profiles and the number of certificate revocation list rotations. The golang version of gRPC requires a complete history of CRL files since the last startup.
+- Memory requirements
+  - A map containing a few details for each profile.
+- Docker dependency
+  - None
+- Build dependency if any
+  - [gNSI](https://github.com/openconfig/gnsi/tree/main) gRPC Network Security Interface
+- Management interfaces - SNMP, CLI, RestAPI, etc.,
+  - [Configuration and management](#configuration-and-management)
+- Is this change specific to any platform? Are there dependencies for platforms to implement anything to make this feature work? If yes, explain in detail and inform community in advance.
+  - Any Platform
+- SAI API requirements, CLI requirements, ConfigDB requirements. Design is covered in following sections.
+
+#### Certz
+
+##### Certz.Rotate
+
+The Certz service definition provides the API to be used for rotating and testing PKI primitives used on network systems. The PKI primitives are grouped by SSL profiles and the service can manage multiple profiles beyond the default.
+
+The Rotate() function is a bidirectional streaming RPC which permits mutating Certificates, Root Certificate Bundles, Certificate Revocation Lists and Authentication Policies of a given profile. In a Rotate() stream it is possible to mutate one or more of the elements, and to send a Finalize message once the in-flight change has been verified to be operational. Failure to send the Finalize message will result in the candidate element being discarded and the original element being used instead.
+
+Certz.Rotate() will permit rotation, and verification of function, of any of the PKI elements. The normal use-case would be to:
+- send a PKI element to a network system as a RotateCertificateRequest.
+- verify that the services which will use the new certificate bundle continue to operate normally.
+- send a FinalizeRequest to finish the rotation process.
+
+The Certz Rotate Flow Diagram
```

Comment:
>Diagram [](http://example.com/codeflow?start=22&length=7)

Is it true that this API is used to rotate only gnmi/gnoi/gnsi used cert? Is it able to rotate other sonic service's cert? like restapi cert.

Reply:
> > Diagram
> 
> Is it true that this API is used to rotate only gnmi/gnoi/gnsi used cert? Is it able to rotate other sonic service's cert? like restapi cert.

The certs managed through gnsi can be used by any service. gnsi supports multi profile for certs as well to manage multiple certs.